docs: add initial guide to node + es modules (#1694)
nuxt/framework: docs/content/2.concepts/4.esm.md

### Aliasing libraries

In some cases, you may also need to manually alias the library to the CJS version, for example:

```js
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
  alias: {
    'sample-library': ['sample-library/dist/sample-library.cjs.js']
  }
})
```

## Library author guide

The good news is that it's relatively simple to fix issues of ESM compatibility. There are really two main options:

1. **You can rename your ESM files to end with `.mjs`.**

   _This is the recommended and simplest approach._ You may have to sort out issues with your library's dependencies, and possibly with your build system, but in most cases this should fix the problem for you. It's also recommended to rename your CJS files to end with `.cjs`, for the greatest explicitness.

1. **You can opt to make your entire library ESM-only**.

   This would mean setting `type: 'module'` in your `package.json` and ensuring that your built library uses ESM syntax. However, you may face issues with your dependencies - and this approach means your library can _only_ be consumed in an ESM context.

### Migration

> Initial step from CJS to ESM, is updating `require` to import:

Author: ```suggestion
The initial step from CJS to ESM is updating any usage of `require` to use `import` instead:
```

::code-group

```js [Before]
module.exports = ...

exports.hello = ...
```

```js [After]
export default ...

export const hello = ...
```

::

::code-group

```js [Before]
const myLib = require('my-lib')
```

```js [After]
import myLib from 'my-lib'
// or
const myLib = await import('my-lib').then(lib => lib.default || lib)
```

::

In ESM Modules, unlike CJS, `require`, `require.resolve`, `__filename` and `__dirname` globals are not available

> and should be replaced with `import()` and `import.meta.filename` when changing from CommonJS to ESM.

Author: ```suggestion
and should be replaced with `import()` and `import.meta.filename`.
```

> You can use [createCommonJS](https://github.com/unjs/mlly#createcommonjs) from `unjs/mlly` to create a CJS compatible context in ESM or use an inline shim:

Author: ```suggestion
You can use [createCommonJS](https://github.com/unjs/mlly#createcommonjs) from `unjs/mlly` to create a CJS compatible context in ESM (or use an inline shim):
```

> 
> 

Author: ```suggestion

```

::code-group

```js [mlly]
import { createCommonJS } from 'mlly'

const { __dirname, __filename, require } = createCommonJS(import.meta.url)
```

```js [manual]
import { fileURLToPath } from 'url'
import { dirname } from 'path'
import { createRequire } from 'module'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const require = createRequire(import.meta.url)
```

::

::code-group

```js [Before]
const someFile = require.resolve('./lib/foo.js')
```

```js [After]
import { resolvePath } from 'mlly'

const someFile = await resolvePath('my-lib', { url: import.meta.url })
```

::

### Best practices

> - Prefer named exports rather than default export. This helps reducing CJS conflicts.

Author: ```suggestion
- Prefer named exports rather than default export. This helps reduce CJS conflicts.
```

> - Avoid depending on Node.js built-ins and CommonJS or Node.js-only dependencies as much as possible to make your library usable in Browsers and Edge Workers without need of Nitro polyfills.

Author: ```suggestion
- Avoid depending on Node.js built-ins and CommonJS or Node.js-only dependencies as much as possible to make your library usable in Browsers and Edge Workers without needing Nitro polyfills.
```

- Use new `exports` field with conditional exports. ([read more](https://nodejs.org/api/packages.html#conditional-exports)).

```json

> {
>   "exports": {
>     "import": "./dist/mymodule.mjs",
>   }
> }

Author: ```suggestion
{
  "exports": {
    ".": {
      "import": "./dist/mymodule.mjs"
    }
  }
}
```
